parser.go

```diff
@@ -146,21 +159,30 @@ func (p *Parser) ParseUnverified(tokenString string, claims Claims) (token *Toke
 	if claimBytes, err = p.DecodeSegment(parts[1]); err != nil {
 		return token, parts, newError("could not base64 decode claim", ErrTokenMalformed, err)
 	}
-	dec := json.NewDecoder(bytes.NewBuffer(claimBytes))
-	if p.useJSONNumber {
-		dec.UseNumber()
-	}
+
 	// JSON Decode.  Special case for map type to avoid weird pointer behavior
-	if c, ok := token.Claims.(MapClaims); ok {
-		err = dec.Decode(&c)
+	mapClaims, isMapClaims := token.Claims.(MapClaims)
+	if p.jsonEncoder != nil {
```

Comment:
Ok this is messy. It seems that maybe the better way would actually be to define an interface not for `Unmarshal` but for `Decode` and use `json.NewDecoder(....)` to fulfil this interface. 

Then we can basically re-use the old code that sets JSON number etc in case `jsonEncoder` (or decoder) is nil.

I am not quite sure how the other JSON libraries work. Do they also have a possible `Decoder` struct?

Reply:
See above. Using Decode(r) instead of Unmarshal should be avoided. Here is a quick suggestions for this code, it looks a bit hacky but does the trick a little bit better I think. I still need to run some numbers. I think this should be faster for most use-cases, where UseNumber is not needed:

```go
// JSON Decode.  Special case for map type to avoid weird pointer behavior
	mapClaims, isMapClaims := token.Claims.(MapClaims)

	var unmarshal func([]byte, any) error

	if p.jsonEncoder == nil {
		if p.useJSONNumber {
			unmarshal = func(data []byte, v any) error {
				decoder := json.NewDecoder(bytes.NewBuffer(claimBytes))
				decoder.UseNumber()
				return decoder.Decode(v)
			}
		} else {
			unmarshal = json.Unmarshal
		}
	} else {
		unmarshal = p.jsonEncoder.Unmarshal
	}

	if isMapClaims {
		err = unmarshal(claimBytes, &mapClaims)
	} else {
		err = unmarshal(claimBytes, &claims)
	}
```